### PR TITLE
[MI-3591] Add transactions in the functions to avoid race conditions

### PR DESCRIPTION
**Pull request by ayusht2810**
#### Summary
Add transactions in the functions to avoid race conditions

#### Ticket Link
#327 
#330

### CODE REVIEW

File: server/command_test.go [hunk 1]
```diff
@@ -681,8 +682,10 @@ func TestExecuteLinkCommand(t *testing.T) {
 				s.On("GetLinkByChannelID", testutils.GetChannelID()).Return(nil, nil).Times(1)
 				s.On("GetLinkByMSTeamsChannelID", testutils.GetTeamsUserID(), testutils.GetChannelID()).Return(nil, nil).Times(1)
 				s.On("GetTokenForMattermostUser", testutils.GetUserID()).Return(&oauth2.Token{}, nil).Times(1)
-				s.On("StoreChannelLink", mock.Anything).Return(nil).Times(1)
-				s.On("SaveChannelSubscription", mock.Anything).Return(nil).Times(1)
+				s.On("StoreChannelLink", mock.AnythingOfType("*storemodels.ChannelLink")).Return(nil).Times(1)
+				s.On("BeginTx").Return(&sql.Tx{}, nil).Times(1)
+				s.On("SaveChannelSubscription", mock.AnythingOfType("storemodels.ChannelSubscription"), &sql.Tx{}).Return(nil).Times(1)
+				s.On("CommitTx", &sql.Tx{}).Return(nil).Times(1)
```
Comment (raghavaggarwal2308) [inline server/command_test.go:688]: Add testcases for errors as well

File: server/monitor/subscriptions.go [hunk 2]
```diff
@@ -218,6 +237,7 @@ func (m *Monitor) GetMSTeamsSubscriptionsMap() (msteamsSubscriptionsMap map[stri
 	msteamsSubscriptionsMap = make(map[string]*msteams.Subscription)
 	for _, msteamsSubscription := range msteamsSubscriptions {
 		if strings.HasPrefix(msteamsSubscription.NotificationURL, m.baseURL) {
+			// m.client.DeleteSubscription(msteamsSubscription.ID)
```
Comment (raghavaggarwal2308) [inline server/monitor/subscriptions.go:240]: need of this?

File: server/store/store.go [hunk 3]
```diff
@@ -984,6 +1069,37 @@ func (s *SQLStore) GetConnectedUsers(page, perPage int) ([]*storemodels.Connecte
 	return connectedUsers, nil
 }
 
+func (s *SQLStore) LockPostByMSTeamsPostID(tx *sql.Tx, messageID string) error {
+	query := s.getQueryBuilder().Select("*").From("msteamssync_posts").Where(sq.Eq{"msTeamsPostID": messageID}).Suffix("FOR UPDATE").RunWith(tx)
+	if _, err := query.Exec(); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (raghavaggarwal2308) [inline server/store/store.go:1078]:
```suggestion
	_, err := query.Exec();
        return err
```
maybe?

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/command_test.go [hunk 4]
```diff
@@ -681,8 +682,96 @@ func TestExecuteLinkCommand(t *testing.T) {
...
+					ServiceSettings: model.ServiceSettings{
+						SiteURL: model.NewString("/"),
+					},
+				}, nil).Times(2)
+				api.On("HasPermissionToChannel", testutils.GetUserID(), testutils.GetChannelID(), model.PermissionManageChannelRoles).Return(true).Times(1)
+				api.On("SendEphemeralPost", testutils.GetUserID(), &model.Post{
+					UserId:    "bot-user-id",
+					ChannelId: testutils.GetChannelID(),
+					Message:   "Please wait while your request is being processed.",
+				}).Return(testutils.GetPost(testutils.GetChannelID(), testutils.GetUserID(), time.Now().UnixMicro())).Times(1)
```
[the remark this answers is not in the document]
Comment (ayusht2810) [reply server/command_test.go]: discussed

File: server/message_hooks.go [hunk 5]
```diff
@@ -220,22 +220,45 @@ func (p *Plugin) SetChatReaction(teamsMessageID, srcUser, channelID, emojiName s
...
+	if txErr = p.store.LockPostByMSTeamsPostID(tx, teamsMessageID); txErr != nil {
+		return txErr
+	}
+
 	if updateRequired {
-		teamsMessage, err = client.SetChatReaction(chatID, teamsMessageID, srcUserID, emoji.Parse(":"+emojiName+":"))
-		if err != nil {
-			p.API.LogError("Error creating post reaction", "error", err.Error())
-			return err
+		teamsMessage, txErr = client.SetChatReaction(chatID, teamsMessageID, srcUserID, emoji.Parse(":"+emojiName+":"))
```
[the remark this answers is not in the document]
Comment (ayusht2810) [reply server/message_hooks.go:247]: discussed

File: server/message_hooks_test.go [hunk 6]
```diff
@@ -927,20 +1174,64 @@ func TestUnsetReaction(t *testing.T) {
...
+			ExpectedMessage: "unable to unset the reaction",
+		},
+		{
+			Name: "UnsetReaction: Unable to rollback database transaction",
+			SetupAPI: func(api *plugintest.API) {
+				api.On("LogError", "Error in removing the reaction", "emojiName", "mockName", "error", ": , unable to unset the reaction").Return(nil).Times(1)
+				api.On("LogWarn", "Unable to rollback database transaction", "error", "unable to rollback database transaction").Return(nil).Times(1)
+			},
+			SetupStore: func(store *storemocks.Store) {
+				store.On("GetPostInfoByMattermostID", testutils.GetID()).Return(&storemodels.PostInfo{}, nil).Times(1)
```
[the remark this answers is not in the document]
Comment (ayusht2810) [reply server/message_hooks_test.go:1193]: discussed

File: server/message_hooks.go [hunk 7]
```diff
@@ -638,31 +728,54 @@ func (p *Plugin) Update(teamID, channelID string, user *model.User, newPost, old
...
+		updatedMessage, txErr = getUpdatedMessage(teamID, channelID, parentID, postInfo.MSTeamsID, client)
+		if txErr != nil {
+			p.API.LogWarn("Error in getting the message from MS Teams", "error", txErr)
+			return txErr
 		}
 	}
 
-	if err = p.store.LinkPosts(storemodels.PostInfo{MattermostID: newPost.Id, MSTeamsChannel: channelID, MSTeamsID: postInfo.MSTeamsID, MSTeamsLastUpdateAt: updatedMessage.LastUpdateAt}); err != nil {
-		p.API.LogWarn("Error updating the msteams/mattermost post link metadata", "error", err)
+	if txErr = p.store.LinkPosts(storemodels.PostInfo{MattermostID: newPost.Id, MSTeamsChannel: channelID, MSTeamsID: postInfo.MSTeamsID, MSTeamsLastUpdateAt: updatedMessage.LastUpdateAt}, tx); txErr != nil {
```
Comment (Kshitij-Katiyar) [inline server/message_hooks.go:777]: Why we are not returning the error here.

File: server/monitor/subscriptions_test.go [hunk 8]
```diff
@@ -536,7 +609,9 @@ func TestMonitorRecreateChannelSubscription(t *testing.T) {
 			setupAPI: func(mockAPI *plugintest.API) {},
 			setupStore: func(store *mocksStore.Store) {
 				store.On("DeleteSubscription", "test-id").Return(nil)
-				store.On("SaveChannelSubscription", storemodels.ChannelSubscription{SubscriptionID: "new-id", TeamID: "team-id", ChannelID: "channel-id", Secret: "webhook-secret", ExpiresOn: newExpiresOn}).Return(nil)
+				store.On("BeginTx").Return(&sql.Tx{}, nil).Times(1)
+				store.On("SaveChannelSubscription", storemodels.ChannelSubscription{SubscriptionID: "new-id", TeamID: "team-id", ChannelID: "channel-id", Secret: "webhook-secret", ExpiresOn: newExpiresOn}, &sql.Tx{}).Return(nil)
+				store.On("CommitTx", &sql.Tx{}).Return(nil).Times(1)
 			},
 		},
 	} {
```
Comment (Kshitij-Katiyar) [inline server/monitor/subscriptions_test.go:617]: The name of the test case should TestRecreateChannelSubscription

File: server/msteams/client.go [hunk 9]
```diff
@@ -737,6 +737,11 @@ func (tc *ClientImpl) UpdateMessage(teamID, channelID, parentID, msgID, message
 		}
 	}
 
+	if updateMessageRequest == nil {
+		tc.logService.Debug("Received nil updateMessageRequest from MS Graph")
+		return nil, errors.New("nil updateMessageRequest")
```
Comment (Kshitij-Katiyar) [inline server/msteams/client.go:742]: Please write user friendly error message

Comment (Kshitij-Katiyar) [reply server/msteams/client.go:742]: Or descriptive log message

File: server/msteams/client.go [hunk 10]
```diff
@@ -738,8 +738,7 @@ func (tc *ClientImpl) UpdateMessage(teamID, channelID, parentID, msgID, message
 	}
 
 	if updateMessageRequest == nil {
-		tc.logService.Debug("Received nil updateMessageRequest from MS Graph")
-		return nil, errors.New("nil updateMessageRequest")
+		return nil, errors.New("Received nil updateMessageRequest from MS Graph")
```
Comment (Kshitij-Katiyar) [inline server/msteams/client.go]: Received nil in updateMessageRequest from MS Graph. Please change this everywhere

**Review by Kshitij-Katiyar — APPROVED** [review]

**Review by manojmalik20 — APPROVED** [review]

**Pull request closed without merging**